File: src/core/store.ts
import Database from 'better-sqlite3'
import type {
  PackageInfo,
  BlobInfo,
  FileRecord,
  MetadataKey,
} from '../types.js'

/** Current schema version */
const SCHEMA_VERSION = '1'

/** Table creation SQL */
const CREATE_TABLES_SQL = `
-- Metadata
CREATE TABLE IF NOT EXISTS metadata (
  key TEXT PRIMARY KEY,
  value TEXT
);

-- Package information
CREATE TABLE IF NOT EXISTS packages (
  id INTEGER PRIMARY KEY AUTOINCREMENT,
  name TEXT NOT NULL,
  version TEXT NOT NULL,
  path TEXT NOT NULL,
  UNIQUE(name, version, path)
);

-- Content-addressable blob storage
CREATE TABLE IF NOT EXISTS blobs (
  hash TEXT PRIMARY KEY,
  content BLOB NOT NULL,
  original_size INTEGER,
  compressed_size INTEGER
);

-- Files per package
CREATE TABLE IF NOT EXISTS files (
  id INTEGER PRIMARY KEY AUTOINCREMENT,
  package_id INTEGER REFERENCES packages(id),
  relative_path TEXT NOT NULL,
  blob_hash TEXT REFERENCES blobs(hash),
  mode INTEGER,
  mtime INTEGER,
  UNIQUE(package_id, relative_path)
);

-- Indexes
CREATE INDEX IF NOT EXISTS idx_files_package ON files(package_id);
CREATE INDEX IF NOT EXISTS idx_files_blob ON files(blob_hash);
`

/**
 * SQLite database store
 */
export class Store {
  /** better-sqlite3 database instance */
  private db: Database.Database

  /**
   * Create Store instance
   * @param dbPath - SQLite database file path
   */
  constructor(dbPath: string) {
    this.db = new Database(dbPath)
    this.db.pragma('journal_mode = WAL')
    this.db.pragma('synchronous = NORMAL')

    this.initSchema()
  }

  /** Initialize schema (create tables and set version) */
  private initSchema(): void {
    this.db.exec(CREATE_TABLES_SQL)

    this.setMetadata('schema_version', SCHEMA_VERSION)
  }

  /**
   * Save metadata
   * @param key - Metadata key
   * @param value - Value to save
   */
  setMetadata(key: MetadataKey, value: string): void {
    const stmt = this.db.prepare(`
      INSERT OR REPLACE INTO metadata (key, value) VALUES (?, ?)
    `)

    stmt.run(key, value)
  }

  /**
   * Get metadata
   * @param key - Metadata key
   * @returns Stored value or null
   */
  getMetadata(key: MetadataKey): string | null {
    const stmt = this.db.prepare(`SELECT value FROM metadata WHERE key = ?`)
    const row = stmt.get(key) as { value: string } | undefined

    return row?.value ?? null
  }

  /**
   * Insert package (returns ID only if already exists)
   * @param pkg - Package information
   * @returns Inserted package ID
   */
  insertPackage(pkg: PackageInfo): number {
    const stmt = this.db.prepare(`
      INSERT INTO packages (name, version, path) VALUES (?, ?, ?)
      ON CONFLICT(name, version, path) DO UPDATE SET name = name
      RETURNING id
    `)
    const result = stmt.get(pkg.name, pkg.version, pkg.path) as { id: number }

    return result.id
  }

  /**
   * Get package by ID
   * @param id - Package ID
   * @returns Package info or null
   */
  getPackageById(id: number): PackageInfo | null {
    const stmt = this.db.prepare(`SELECT * FROM packages WHERE id = ?`)
    const row = stmt.get(id) as
      | { id: number; name: string; version: string; path: string }
      | undefined

    if (!row) return null

    return {
      id: row.id,
      name: row.name,
      version: row.version,
      path: row.path,
    }
  }

  /**
   * Get all packages
   * @returns All package list
   */
  getAllPackages(): PackageInfo[] {
    const stmt = this.db.prepare(`SELECT * FROM packages`)
    const rows = stmt.all() as Array<{
      id: number
      name: string
      version: string
      path: string
    }>

    return rows.map((row) => ({
      id: row.id,
      name: row.name,
      version: row.version,
      path: row.path,
    }))
  }

  /**
   * Check if blob exists
   * @param hash - Blob hash
   * @returns Whether blob exists
   */
  hasBlob(hash: string): boolean {
    const stmt = this.db.prepare(`SELECT 1 FROM blobs WHERE hash = ?`)

    return stmt.get(hash) !== undefined
  }

  /**
   * Insert blob (ignore duplicates)
   * @param blob - Blob information
   */
  insertBlob(blob: BlobInfo): void {
    const stmt = this.db.prepare(`
      INSERT OR IGNORE INTO blobs (hash, content, original_size, compressed_size)
      VALUES (?, ?, ?, ?)
    `)

    stmt.run(blob.hash, blob.content, blob.originalSize, blob.compressedSize)
  }

  /**
   * Get blob content by hash
   * @param hash - Blob hash
   * @returns Compressed blob content or null
   */
  getBlob(hash: string): Buffer | null {
    const stmt = this.db.prepare(`SELECT content FROM blobs WHERE hash = ?`)
    const row = stmt.get(hash) as { content: Buffer } | undefined

    return row?.content ?? null
  }

  /**
   * Get blob storage statistics
   * @returns Blob count and size statistics
   */
  getBlobStats(): {
    totalBlobs: number
    totalOriginalSize: number
    totalCompressedSize: number
  } {
    const stmt = this.db.prepare(`
      SELECT
        COUNT(*) as count,
        COALESCE(SUM(original_size), 0) as original,
        COALESCE(SUM(compressed_size), 0) as compressed
      FROM blobs
    `)
    const row = stmt.get() as {
      count: number
      original: number
      compressed: number
    }

    return {
      totalBlobs: row.count,
      totalOriginalSize: row.original,
      totalCompressedSize: row.compressed,
    }
  }

  /**
   * Insert file record (update on conflict)
   * @param file - File record information
   */
  insertFile(file: FileRecord): void {
    const stmt = this.db.prepare(`
      INSERT INTO files (package_id, relative_path, blob_hash, mode, mtime)
      VALUES (?, ?, ?, ?, ?)
      ON CONFLICT(package_id, relative_path) DO UPDATE SET
        blob_hash = excluded.blob_hash,
        mode = excluded.mode,
        mtime = excluded.mtime
    `)

    stmt.run(
      file.packageId,
      file.relativePath,
      file.blobHash,
      file.mode,
      file.mtime
    )
  }

  /**
   * Get files by package ID
   * @param packageId - Package ID
   * @returns File records for the package
   */
  getFilesByPackage(packageId: number): FileRecord[] {
    const stmt = this.db.prepare(`
      SELECT id, package_id, relative_path, blob_hash, mode, mtime
      FROM files WHERE package_id = ?
    `)
    const rows = stmt.all(packageId) as Array<{
      id: number
      package_id: number
      relative_path: string
      blob_hash: string
      mode: number
      mtime: number
    }>

    return rows.map((row) => ({
      id: row.id,
      packageId: row.package_id,
      relativePath: row.relative_path,
      blobHash: row.blob_hash,
      mode: row.mode,
      mtime: row.mtime,
    }))
  }

  /**
   * Get all files (with package path)
   * @returns All file records
   */
  getAllFiles(): Array<FileRecord & { packagePath: string }> {
    const stmt = this.db.prepare(`
      SELECT f.id, f.package_id, f.relative_path, f.blob_hash, f.mode, f.mtime, p.path as package_path
      FROM files f
      JOIN packages p ON f.package_id = p.id
    `)
    const rows = stmt.all() as Array<{
      id: number
      package_id: number
      relative_path: string
      blob_hash: string
      mode: number
      mtime: number
      package_path: string
    }>

    return rows.map((row) => ({
      id: row.id,
      packageId: row.package_id,
      relativePath: row.relative_path,
      blobHash: row.blob_hash,
      mode: row.mode,
      mtime: row.mtime,
      packagePath: row.package_path,
    }))
  }

  /**
   * Get total file count
   * @returns Total number of files
   */
  getTotalFileCount(): number {
    const stmt = this.db.prepare(`SELECT COUNT(*) as count FROM files`)
    const row = stmt.get() as { count: number }

    return row.count
  }

  /**
   * Execute work within transaction
   * @param fn - Function to execute within transaction
   * @returns Function execution result
   */
  transaction<T>(fn: () => T): T {
    return this.db.transaction(fn)()
  }

  /**
   * Get prepared statement for bulk file insertion
   * @returns Prepared statement for file insertion
   */
  prepareInsertFile(): Database.Statement {
    return this.db.prepare(`
      INSERT INTO files (package_id, relative_path, blob_hash, mode, mtime)
      VALUES (?, ?, ?, ?, ?)
      ON CONFLICT(package_id, relative_path) DO UPDATE SET
        blob_hash = excluded.blob_hash,
        mode = excluded.mode,
        mtime = excluded.mtime
    `)
  }

  /**
   * Get prepared statement for bulk blob insertion
   * @returns Prepared statement for blob insertion
   */
  prepareInsertBlob(): Database.Statement {
    return this.db.prepare(`
      INSERT OR IGNORE INTO blobs (hash, content, original_size, compressed_size)
      VALUES (?, ?, ?, ?)
    `)
  }

  /** Close database connection */
  close(): void {
    this.db.close()
  }
}

File: src/core/store.rs
use anyhow::{anyhow, bail, Context, Result};
use rusqlite::{params, Connection, OpenFlags, Transaction};

use crate::types::{BlobStats, FileRecord, FileRecordWithPath, LinkEntry, PackageInfo};

const SCHEMA_VERSION: &str = "3";

const CREATE_TABLES_SQL: &str = "
CREATE TABLE IF NOT EXISTS metadata (
  key TEXT PRIMARY KEY,
  value TEXT
);

CREATE TABLE IF NOT EXISTS packages (
  id INTEGER PRIMARY KEY AUTOINCREMENT,
  name TEXT NOT NULL,
  version TEXT NOT NULL,
  path TEXT NOT NULL,
  UNIQUE(name, version, path)
);

CREATE TABLE IF NOT EXISTS blobs (
  hash BLOB PRIMARY KEY,
  content BLOB NOT NULL,
  original_size INTEGER,
  compressed_size INTEGER
);

CREATE TABLE IF NOT EXISTS files (
  id INTEGER PRIMARY KEY AUTOINCREMENT,
  package_id INTEGER REFERENCES packages(id),
  relative_path TEXT NOT NULL,
  blob_hash BLOB REFERENCES blobs(hash),
  mode INTEGER,
  mtime INTEGER,
  UNIQUE(package_id, relative_path)
);

CREATE TABLE IF NOT EXISTS links (
  path TEXT PRIMARY KEY,
  target TEXT NOT NULL
);

CREATE TABLE IF NOT EXISTS dirs (
  path TEXT PRIMARY KEY
);

CREATE INDEX IF NOT EXISTS idx_files_package ON files(package_id);
CREATE INDEX IF NOT EXISTS idx_files_blob ON files(blob_hash);
";

pub struct Store {
    conn: Connection,
}

pub fn insert_package(tx: &Transaction, pkg: &PackageInfo) -> Result<i64> {
    let mut stmt = tx.prepare_cached(
        "INSERT INTO packages (name, version, path) VALUES (?1, ?2, ?3)
         ON CONFLICT(name, version, path) DO UPDATE SET name = name
         RETURNING id",
    )?;
    let id: i64 = stmt.query_row(params![pkg.name, pkg.version, pkg.path], |row| row.get(0))?;
    Ok(id)
}

pub fn insert_blob(
    tx: &Transaction,
    hash: &[u8],
    content: &[u8],
    original_size: u64,
) -> Result<()> {
    let mut stmt = tx.prepare_cached(
        "INSERT OR IGNORE INTO blobs (hash, content, original_size, compressed_size)
         VALUES (?1, ?2, ?3, ?4)",
    )?;
    stmt.execute(params![hash, content, original_size, content.len() as u64])?;
    Ok(())
}

pub fn insert_file(
    tx: &Transaction,
    package_id: i64,
    relative_path: &str,
    blob_hash: &[u8],
    mode: u32,
    mtime: i64,
) -> Result<()> {
    let mut stmt = tx.prepare_cached(
        "INSERT INTO files (package_id, relative_path, blob_hash, mode, mtime)
         VALUES (?1, ?2, ?3, ?4, ?5)
         ON CONFLICT(package_id, relative_path) DO UPDATE SET
           blob_hash = excluded.blob_hash,
           mode = excluded.mode,
           mtime = excluded.mtime",
    )?;
    stmt.execute(params![package_id, relative_path, blob_hash, mode, mtime])?;
    Ok(())
}

pub fn insert_link(tx: &Transaction, link: &LinkEntry) -> Result<()> {
    let mut stmt =
        tx.prepare_cached("INSERT OR REPLACE INTO links (path, target) VALUES (?1, ?2)")?;
    stmt.execute(params![link.path, link.target])?;
    Ok(())
}

pub fn insert_dir(tx: &Transaction, path: &str) -> Result<()> {
    let mut stmt = tx.prepare_cached("INSERT OR REPLACE INTO dirs (path) VALUES (?1)")?;
    stmt.execute(params![path])?;
    Ok(())
}

impl Store {
    pub fn create(db_path: &str) -> Result<Self> {
        let conn = Connection::open(db_path)?;
        conn.pragma_update(None, "journal_mode", "WAL")?;
        conn.pragma_update(None, "synchronous", "NORMAL")?;

        conn.execute_batch(CREATE_TABLES_SQL)?;

        let store = Store { conn };
        store.set_metadata("schema_version", SCHEMA_VERSION)?;

        Ok(store)
    }

    pub fn finalize(self) -> Result<()> {
        self.conn.pragma_update(None, "journal_mode", "DELETE")?;
        Ok(())
    }

    pub fn open_readonly(db_path: &str) -> Result<Self> {
        let conn = Connection::open_with_flags(db_path, OpenFlags::SQLITE_OPEN_READ_ONLY)?;
        let store = Store { conn };

        let version = store
            .get_metadata("schema_version")
            .with_context(|| format!("not a mohyung database: {}", db_path))?
            .ok_or_else(|| anyhow!("not a mohyung database: {}", db_path))?;

        if version != SCHEMA_VERSION {
            bail!(
                "unsupported schema version {} (this build supports {}): {}",
                version,
                SCHEMA_VERSION,
                db_path
            );
        }

        Ok(store)
    }

    pub fn set_metadata(&self, key: &str, value: &str) -> Result<()> {
        self.conn.execute(
            "INSERT OR REPLACE INTO metadata (key, value) VALUES (?1, ?2)",
            params![key, value],
        )?;
        Ok(())
    }

    pub fn get_metadata(&self, key: &str) -> Result<Option<String>> {
        let mut stmt = self
            .conn
            .prepare("SELECT value FROM metadata WHERE key = ?1")?;
        let result = stmt
            .query_row(params![key], |row| row.get::<_, String>(0))
            .ok();
        Ok(result)
    }

    pub fn get_blob(&self, hash: &[u8]) -> Result<Option<Vec<u8>>> {
        let mut stmt = self
            .conn
            .prepare_cached("SELECT content FROM blobs WHERE hash = ?1")?;
        let result = stmt
            .query_row(params![hash], |row| row.get::<_, Vec<u8>>(0))
            .ok();
        Ok(result)
    }

    pub fn get_blob_stats(&self) -> Result<BlobStats> {
        let mut stmt = self.conn.prepare(
            "SELECT COUNT(*) as count,
                    COALESCE(SUM(original_size), 0) as original,
                    COALESCE(SUM(compressed_size), 0) as compressed
             FROM blobs",
        )?;
        let stats = stmt.query_row([], |row| {
            Ok(BlobStats {
                total_blobs: row.get::<_, i64>(0)? as usize,
                total_original_size: row.get::<_, i64>(1)? as u64,
                total_compressed_size: row.get::<_, i64>(2)? as u64,
            })
        })?;
        Ok(stats)
    }

    pub fn get_all_files(&self) -> Result<Vec<FileRecordWithPath>> {
        let mut stmt = self.conn.prepare(
            "SELECT f.id, f.package_id, f.relative_path, f.blob_hash, f.mode, f.mtime, p.path as package_path
             FROM files f
             JOIN packages p ON f.package_id = p.id
             ORDER BY f.blob_hash",
        )?;
        let rows = stmt.query_map([], |row| {
            Ok(FileRecordWithPath {
                record: FileRecord {
                    id: Some(row.get::<_, i64>(0)?),
                    package_id: row.get(1)?,
                    relative_path: row.get(2)?,
                    blob_hash: row.get(3)?,
                    mode: row.get::<_, u32>(4)?,
                    mtime: row.get(5)?,
                },
                package_path: row.get(6)?,
            })
        })?;

        let mut files = Vec::new();
        for row in rows {
            files.push(row?);
        }
        Ok(files)
    }

    pub fn get_all_links(&self) -> Result<Vec<LinkEntry>> {
        let mut stmt = self
            .conn
            .prepare("SELECT path, target FROM links ORDER BY path")?;
        let rows = stmt.query_map([], |row| {
            Ok(LinkEntry {
                path: row.get(0)?,
                target: row.get(1)?,
            })
        })?;

        let mut links = Vec::new();
        for row in rows {
            links.push(row?);
        }
        Ok(links)
    }

    pub fn get_all_dirs(&self) -> Result<Vec<String>> {
        let mut stmt = self.conn.prepare("SELECT path FROM dirs ORDER BY path")?;
        let rows = stmt.query_map([], |row| row.get::<_, String>(0))?;

        let mut dirs = Vec::new();
        for row in rows {
            dirs.push(row?);
        }
        Ok(dirs)
    }

    pub fn get_total_file_count(&self) -> Result<usize> {
        let mut stmt = self.conn.prepare("SELECT COUNT(*) FROM files")?;
        let count: i64 = stmt.query_row([], |row| row.get(0))?;
        Ok(count as usize)
    }

    pub fn transaction<F, T>(&mut self, f: F) -> Result<T>
    where
        F: FnOnce(&Transaction) -> Result<T>,
    {
        let tx = self.conn.transaction()?;
        let result = f(&tx)?;
        tx.commit()?;
        Ok(result)
    }
}
